plugin/rewrite/name.go

```diff
@@ -11,54 +11,60 @@ import (
 	"github.com/coredns/coredns/request"
 )
 
-type exactNameRule struct {
-	NextAction string
-	From       string
-	To         string
-	ResponseRule
+const (
+	// ExactMatch matches only on exact match of the name in the question section of a request
+	ExactMatch = "exact"
+	// PrefixMatch matches when the name begins with the matching string
+	PrefixMatch = "prefix"
+	// SuffixMatch matches when the name ends with the matching string
+	SuffixMatch = "suffix"
+	// SubstringMatch matches on partial match of the name in the question section of a request
+	SubstringMatch = "substring"
+	// RegexMatch matches when the name in the question section of a request matches a regular expression
+	RegexMatch = "regex"
+)
+
+type nameRule struct {
+	nextAction   string
+	responseRule ResponseRule
+
+	rewriter
+}
+
+// rewriter is an internal interface that handles the specific rewrite behavior
+// of each rewrite rule type
+type rewriter interface {
+	Rewrite(ctx context.Context, state request.Request) Result
 }
 
-type prefixNameRule struct {
-	NextAction  string
+type exactRewriter struct {
+	From string
+	To   string
+}
+
+type prefixRewriter struct {
 	Prefix      string
 	Replacement string
 }
 
-type suffixNameRule struct {
-	NextAction  string
+type suffixRewriter struct {
 	Suffix      string
 	Replacement string
 }
 
-type substringNameRule struct {
-	NextAction  string
+type substringRewriter struct {
 	Substring   string
 	Replacement string
 }
 
-type regexNameRule struct {
-	NextAction  string
+type regexRewriter struct {
 	Pattern     *regexp.Regexp
 	Replacement string
-	ResponseRule
 }
 
-const (
-	// ExactMatch matches only on exact match of the name in the question section of a request
-	ExactMatch = "exact"
-	// PrefixMatch matches when the name begins with the matching string
-	PrefixMatch = "prefix"
-	// SuffixMatch matches when the name ends with the matching string
-	SuffixMatch = "suffix"
-	// SubstringMatch matches on partial match of the name in the question section of a request
-	SubstringMatch = "substring"
-	// RegexMatch matches when the name in the question section of a request matches a regular expression
-	RegexMatch = "regex"
-)
-
 // Rewrite rewrites the current request based upon exact match of the name
 // in the question section of the request.
-func (rule *exactNameRule) Rewrite(ctx context.Context, state request.Request) Result {
+func (rule exactRewriter) Rewrite(ctx context.Context, state request.Request) Result {
```

Comment:
Per the go FAQ: https://golang.org/doc/faq#methods_on_values_or_pointers

>  For types such as basic types, slices, and small structs, a value receiver is very cheap so unless the semantics of the method requires a pointer, a value receiver is efficient and clear. 

The semantics of Rewrite do not require a pointer receiver, so it's more clear to use a value receiver, as the Go language authors recommend.

I'm happy to split this into its own PR if you'd like since it is pretty independent of this change.